service/topology_coordinator.cc

```diff
@@ -850,6 +853,90 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+            // TODO: Exclude nodes which don't contain replicas of the table we are truncating
+            guard = co_await global_tablet_token_metadata_barrier(std::move(guard));
+
+            while (true) {
+                topology& topo = _topo_sm._topology;
+                const utils::UUID& global_request_id = topo.global_request_id.value();
+                const table_id& table_id = topo.truncate_table_id.value();
+                const std::chrono::milliseconds timeout = topo.truncate_table_timeout_ms.value();
+                lw_shared_ptr<replica::table> table = _db.get_tables_metadata().get_table_if_exists(table_id);
+
+                std::optional<sstring> error;
+                if (table) {
+                    const sstring& ks_name = table->schema()->ks_name();
+                    const sstring& cf_name = table->schema()->cf_name();
+
+                    rtlogger.info("Performing TRUNCATE TABLE global topology request for {}.{} with timeout {}", ks_name, cf_name, timeout);
+
+                    const locator::tablet_map& tmap = get_token_metadata_ptr()->tablets().get_tablet_map(table_id);
+
+                    // Collect the IPs of the hosts with replicas
+                    std::unordered_set<gms::inet_address> replica_endpoints;
+                    co_await tmap.for_each_tablet([&] (locator::tablet_id tid, const locator::tablet_info& tinfo) {
+                        for (const locator::tablet_replica& replica: tinfo.replicas) {
+                            const std::set<gms::inet_address> addrs = _gossiper.get_nodes_with_host_id(replica.host);
+                            replica_endpoints.insert(addrs.begin(), addrs.end());
+                        }
+                        return make_ready_future<>();
+                    });
+
+                    // Check if all the nodes with replicas are alive
+                    const std::set<gms::inet_address> live_nodes = _gossiper.get_live_members();
+                    const size_t live_replicas = std::count_if(replica_endpoints.begin(), replica_endpoints.end(),
+                                                                [&live_nodes] (const gms::inet_address& addr) {
+                                                                    return live_nodes.contains(addr);
+                                                                });
+                    if (live_replicas == replica_endpoints.size()) {
+                        try {
+                            // Send the RPC to all replicas
+                            const netw::messaging_service::clock_type::time_point timeout_deadline = lowres_clock::now() + timeout;
+                            const service::frozen_topology_guard frozen_guard { _topo_sm._topology.session };
+                            co_await coroutine::parallel_for_each(replica_endpoints, [&] (const gms::inet_address& ep) -> future<> {
+                                netw::messaging_service::msg_addr addr(ep);
```

Comment:
I think it's necessary to freeze the topology and then perform the truncate in two separate RPC calls. Otherwise the following may happen:
1) Tablet starts migrating from A to B.
2) B's topology is frozen
3) B's truncated
4) Tablet data arrives from A
5)  A's topology is frozen
6) A's truncated.

First we should freeze the topology on all participants, and then, knowing no migrations are taking place, we can safely truncate.

Also consider this as a precaution measure: what happens if the topology freeze fails? Will the other nodes be aware of it? Most likely no. And it's safer to not even start truncate inthis case and return an error to the client. 

@tgrabiec @gleb-cloudius please comment.

Reply:
Also please consider what happens here if TRUNCATE  map/reduce here is racing with DROP TABLE.

Reply:
> I think it's necessary to freeze the topology and then perform the truncate in two separate RPC calls. 

I am not sure why this is necessary. We are running TRUNCATE from the topology coordinator, and because of that, no other topology operation can occur concurrently. AFAIK, load balancing migrations will wait for TRUNCATE to complete, and TRUNCATE will wait for the ongoing migration to complete. Am I missing something?

There might be a misunderstanding; the purpose of the `frozen_topology_guard` is to avoid situations where are executing stale RPC calls after the topology coordinator has moved on. The mechanism is described [here](https://github.com/scylladb/scylladb/blob/master/docs/dev/topology-over-raft.md#topology-guards)

Reply:
I do not assume there is only one topology coordinator in the cluster at a given point in time.  A leader may change, and the group0_guard held by the old coordinator will not be effective. The new leader may start from a new raft log position. Yet you're doing an RPC, not a Raft operation, so your RPC may arrive out of order, This is why it's important to perform any side effects as part of applying the raft log, not as an independent rpc.

Reply:
migrations don't run as long as the truncate request is active.

If there is topology coordinator failover then the new coordinator will retry the truncate request and then clear the session. Before any migration starts later, we execute topology barrier, which will wait for any stale RPCs running with now-stale session. So I think we're safe, but.. this relies on the fact that other ops execute the barrier as the first step. It would be more graceful to execute the barrier here, after clearing the session but before completing the request. This way we would have an invariant that if there is no request, there are no stale operations running on behalf of that request.